<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <!-- DVS drivers -->
  <include file="$(find edvs_ros_driver)/launch/edvs-stereo.launch" />

  <!-- visualization -->
  <node name="dvs_renderer_left" pkg="dvs_renderer" type="dvs_renderer" ns="dvs_left">
    <param name="display_method" value="grayscale"/>
  </node>

  <node name="dvs_renderer_right" pkg="dvs_renderer" type="dvs_renderer" ns="dvs_right">
    <param name="display_method" value="grayscale"/>
  </node>

  <!-- calib node -->
  <node name="dvs_calibration" pkg="dvs_calibration" type="stereo" output="screen">
    <remap from="camera_info_left" to="/dvs_left/camera_info" />
    <remap from="events_left" to="/dvs_left/events" />
    <remap from="camera_info_right" to="/dvs_right/camera_info" />
    <remap from="events_right" to="/dvs_right/events" />
    <remap from="set_camera_info_left" to="/dvs_left/set_camera_info" />
    <remap from="set_camera_info_right" to="/dvs_right/set_camera_info" />
    
    <!-- in order to simply take all detected patterns -->
    <!--
    <remap from="image_object_points_left" to="/dvs_calibration/detected_points_left_or_single" />
	-->
	<!-- use output of picker tool for calibration input -->
	<remap from="image_object_points_left" to="/dvs_left/out/image_object_points" />
	<remap from="image_object_points_right" to="/dvs_right/out/image_object_points" />

    <param name="blinking_time_us" value="1000" />
    <param name="blinking_time_tolerance_us" value="100" />
    <!-- enough transitions to start to try to find a pattern -->
    <param name="enough_transitions_threshold" value="20" />
    <!-- minimum transitions of a pixel in order to get used for the point detection -->
    <param name="minimum_transitions_threshold" value="1" />
    <!-- minimum led mass of a cluser in order to get accepted -->
    <param name="minimum_led_mass" value="1" />
    <!-- seconds after which the transition map will be reset-->
    <param name="pattern_search_timeout" value="1"/>
    
    <param name="dots_w" value="4"/>
    <param name="dots_h" value="3"/>
    <!-- distance of the dots in m -->
    <param name="dot_distance" value="0.04"/>
  </node>

  <node name="rqt_gui" pkg="rqt_gui" type="rqt_gui" args="--perspective-file $(find dvs_calibration_gui)/rqt/StereoCalibration.perspective" />


  <!-- pattern picker to confirm detections -->
  <node name="pattern_picker" pkg="dvs_calibration" type="pick.py" output="screen" ns="dvs_left">
	  <remap from="in/image_object_points" to="/dvs_calibration/detected_points_left_or_single" />
	  <remap from="in/image_pattern" to="/dvs_calibration/detected_points_left_or_single_pattern" />
  </node>
  
    <!-- pattern picker to confirm detections -->
  <node name="pattern_picker" pkg="dvs_calibration" type="pick.py" output="screen" ns="dvs_right">
	  <remap from="in/image_object_points" to="/dvs_calibration/detected_points_right" />
	  <remap from="in/image_pattern" to="/dvs_calibration/detected_points_right_pattern" />
	  <param name="stereoslave" value="1" />
	  <remap from="in/ignore" to="/dvs_left/out/ignore" />
	  <remap from="in/take" to="/dvs_left/out/take" />
  </node>
</launch>
